internal/overlord/logstate/loki.go

```diff
@@ -0,0 +1,250 @@
+//		if err := f.parser.Err(); err != nil {
+//			logger.Noticef(
+//				"Forwarding logs to target %q: cannot parse log for service %q: %v",
+//				f.targetName, f.serviceName, err)
+//			// TODO: create a new parser?
+//		}
+//		if !f.buffer.ReadyToFlush() {
+//			f.flushBuffer()
+//		}
+//	}
+//
+//	// Forwarder stopped or service exited - flush remaining logs
+//	if len(f.buffer.messages) > 0 {
+//		f.flushBuffer()
+//	}
+//}
+
+func (f *lokiForwarder) Stop() {
+	f.stopCh <- struct{}{}
+}
+
+func (f *lokiForwarder) flushBuffer() {
+	err := f.client.Send(f.buffer.Request())
+	if err != nil {
+		logger.Noticef(
+			"Forwarding logs to target %q: cannot send logs for service %q: %v",
+			f.targetName, f.serviceName, err)
+		// TODO: retry?
+	}
+
+	f.buffer.Reset()
+}
+
+type lokiBuffer struct {
+	labels map[string]string
+
+	buffer  *bytes.Buffer
+	encoder *json.Encoder
+
+	messages    int
+	maxMessages int
+	deadline    *time.Time
+	delay       time.Duration
+}
+
+func newLokiBuffer(labels map[string]string, maxMessages int, delay time.Duration) *lokiBuffer {
+	buf := bytes.NewBuffer(make([]byte, 0, 4096))
+
+	b := &lokiBuffer{
+		labels:      labels,
+		buffer:      buf,
+		encoder:     json.NewEncoder(buf),
+		maxMessages: maxMessages,
+		delay:       delay,
+	}
+	b.Reset()
+	return b
+}
+
+func (b *lokiBuffer) Reset() {
+	b.messages = 0
+	b.deadline = nil
+	b.buffer.Reset()
+
+	// Write start of request
+	b.buffer.WriteString(`{"streams":[{"stream":`)
+	b.encoder.Encode(b.labels)
+	b.buffer.WriteString(`,"values":[`)
+}
+
+func (b *lokiBuffer) WriteEntry(entry servicelog.Entry) {
+	// Write in format
+	//    ["<unix epoch in nanoseconds>","<log line>"]
+	b.buffer.WriteString("[")
+	timestamp := strconv.FormatInt(entry.Time.UnixNano(), 10)
+	b.encoder.Encode(timestamp)
+	b.buffer.WriteString(",")
+	message := strings.Trim(entry.Message, "\n")
+	b.encoder.Encode(message)
+	b.buffer.WriteString("],")
+
+	b.messages++
+	if b.deadline == nil {
+		// Set write deadline
+		deadline := time.Now().Add(b.delay)
+		b.deadline = &deadline
+	}
+}
+
+func (b *lokiBuffer) ReadyToFlush() bool {
+	return (b.messages >= b.maxMessages) ||
+		(b.deadline != nil && time.Now().After(*b.deadline))
+	// TODO: flush when buffer length is past a certain threshold?
+}
+
+// Prepares a JSON request for sending to Loki.
+func (b *lokiBuffer) Request() io.Reader {
+	// Need to remove last ',' in buffer
+	b.buffer.Truncate(b.buffer.Len() - 1)
+	// Write end of request
+	b.buffer.WriteString("]}]}")
+
+	return b.buffer
+}
+
+const (
+	lokiPushAPI = "/loki/api/v1/push"
+)
+
+// lokiClient handles sending requests to Loki.
+type lokiClient struct {
+	client *http.Client
+	url    string
+}
+
+func newLokiClient(location string) *lokiClient {
+	return &lokiClient{
+		client: &http.Client{
+			Timeout: 10 * time.Second,
+		},
+		url: location,
+	}
+}
+
+func (c *lokiClient) Send(body io.Reader) error {
+	resp, err := c.client.Post(c.url+lokiPushAPI, "application/json", body)
```

Comment:
This is dangerous as mention in previous PR comments.

What you don't know here is the contents of the URL and what this environment may be doing. What if all of my Loki's are behind a load balancer and the URL I give you has a query param in it to help control the logic in my load balancer?

We need to use the appropriate GO libs here to modify the URL path and not just append to a string hope for the best. This can be done once on construction.

Reply:
You're still making assumptions about the user's Loki setup. Ben and I have agreed that it's safer to just ask the full API endpoint, e.g. `https://232.12.67.100:3100/loki/api/v1/push`. This is what Promtail does.